File: rewrite-kotlin/src/test/java/org/openrewrite/kotlin/format/WrappingAndBracesTest.java
/*
 * Copyright 2023 the original author or authors.
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * https://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.openrewrite.kotlin.format;

import org.junit.jupiter.api.Test;
import org.openrewrite.DocumentExample;
import org.openrewrite.Issue;
import org.openrewrite.Tree;
import org.openrewrite.kotlin.KotlinParser;
import org.openrewrite.kotlin.style.IntelliJ;
import org.openrewrite.kotlin.style.SpacesStyle;
import org.openrewrite.kotlin.style.WrappingAndBracesStyle;
import org.openrewrite.style.NamedStyles;
import org.openrewrite.test.RecipeSpec;
import org.openrewrite.test.RewriteTest;

import java.util.List;
import java.util.function.Consumer;
import java.util.function.UnaryOperator;

import static java.util.Collections.emptySet;
import static java.util.Collections.singletonList;
import static org.openrewrite.kotlin.Assertions.kotlin;
import static org.openrewrite.test.RewriteTest.toRecipe;

@SuppressWarnings("All")
class WrappingAndBracesTest implements RewriteTest {

    @Override
    public void defaults(RecipeSpec spec) {
        spec.recipe(toRecipe(() -> new WrappingAndBracesVisitor<>(IntelliJ.wrappingAndBraces())));
    }

    @DocumentExample
    @SuppressWarnings({"ClassInitializerMayBeStatic", "ReassignedVariable", "UnusedAssignment"})
    @Test
    void blockLevelStatements() {
        rewriteRun(
          kotlin(
            """
              class Test {
                  init {        var n: Int = 0
                      n++
                  }
              }
              """,
            """
              class Test {
                  init {
                      var n: Int = 0
                      n++
                  }
              }
              """
          )
        );
    }

    private static Consumer<RecipeSpec> wrappingAndBraces(UnaryOperator<SpacesStyle> spaces,
                                                          UnaryOperator<WrappingAndBracesStyle> wrapping) {
        return spec -> spec
          .recipes(
            toRecipe(() -> new WrappingAndBracesVisitor<>(wrapping.apply(IntelliJ.wrappingAndBraces()))),
            toRecipe(() -> new TabsAndIndentsVisitor<>(IntelliJ.tabsAndIndents(), IntelliJ.wrappingAndBraces())),
            toRecipe(() -> new SpacesVisitor<>(spaces.apply(IntelliJ.spaces())))
          )
          .parser(KotlinParser.builder().styles(singletonList(
            new NamedStyles(
              Tree.randomId(), "test", "test", "test", emptySet(),
              List.of(
                spaces.apply(IntelliJ.spaces()),
                wrapping.apply(IntelliJ.wrappingAndBraces())
              )
            )
          )));
    }

    @Test
    void classConstructor() {
        rewriteRun(
          kotlin(
            """
              class A   (
                      val type: Int = 1
              ) {
                   var a = 2
              }
              """
          )
        );
    }

    @Test
    void blockEndOnOwnLine() {
        rewriteRun(
          kotlin(
            """
              class Test {
                  val n: Int = 0}
              """,
            """
              class Test {
                  val n: Int = 0
              }
              """
          )
        );
    }

    @Test
    void annotatedMethod() {
        rewriteRun(
          kotlin(
            """
              class Test {
                  @Suppress("ALL")   fun method(): Any {
                      return Any()
                  }
              }
              """,
            """
              class Test {
                  @Suppress("ALL")
                 fun method(): Any {
                      return Any()
                  }
              }
              """
          )
        );
    }

    @Test
    void leadingAnnotationNewLine() {
        rewriteRun(
          kotlin(
            """
              class Test {   @Suppress("ALL")   fun method(): Any {
                      return Any()
                  }
              }
              """,
            """
              class Test {
                 @Suppress("ALL")
                 fun method(): Any {
                      return Any()
                  }
              }
              """
          )
        );
    }

    @Test
    void annotatedMethodWithPublicModifier() {
        rewriteRun(
          kotlin(
            """
              class Test {
                  @Suppress("ALL")   public fun method(): Any {
                      return Any()
                  }
              }
              """,
            """
              class Test {
                  @Suppress("ALL")
                 public fun method(): Any {
                      return Any()
                  }
              }
              """
          )
        );
    }

    @Test
    void annotatedMethodWithFinalModifier() {
        rewriteRun(
          kotlin(
            """
              class Test {
                  @Suppress("ALL")   final fun method(): Any {
                      return Any()
                  }
              }
              """,
            """
              class Test {
                  @Suppress("ALL")
                 final fun method(): Any {
                      return Any()
                  }
              }
              """
          )
        );
    }

    @Test
    void annotatedMethodWithModifiers() {
        rewriteRun(
          kotlin(
            """
              class Test {
                  @Suppress("ALL")   public final fun method(): Any {
                      return Any()
                  }
              }
              """,
            """
              class Test {
                  @Suppress("ALL")
                 public final fun method(): Any {
                      return Any()
                  }
              }
              """
          )
        );
    }

    @Test
    void annotatedMethodWithTypeParameter() {
        rewriteRun(
          kotlin(
            """
              class Test {
                  @Suppress("ALL")   fun <T> method(): T? {
                      return null
                  }
              }
              """,
            """
              class Test {
                  @Suppress("ALL")
                 fun <T> method(): T? {
                      return null
                  }
              }
              """
          )
        );
    }

    @Test
    void multipleAnnotatedMethod() {
        rewriteRun(
          kotlin(
            """
              class Test {
                  @Suppress("ALL") @Deprecated("", ReplaceWith("Any()"))   fun method(): Any {
                      return Any()
                  }
              }
              """,
            """
              class Test {
                  @Suppress("ALL")
               @Deprecated("", ReplaceWith("Any()"))
                 fun method(): Any {
                      return Any()
                  }
              }
              """
          )
        );
    }

    @Test
    void annotatedConstructor() {
        rewriteRun(
          kotlin(
            """
              class Test {
                  @Suppress("ALL") @Deprecated("",ReplaceWith("Any()")) constructor() {
                  }
              }
              """,
            """
              class Test {
                  @Suppress("ALL")
               @Deprecated("",ReplaceWith("Any()"))
               constructor() {
                  }
              }
              """
          )
        );
    }

    @Test
    void annotatedClassDecl() {
        rewriteRun(
          kotlin(
            """
              @Suppress("ALL")   class Test {
              }
              """,
            """
              @Suppress("ALL")
                 class Test {
              }
              """
          )
        );
    }

    @Test
    void annotatedClassDeclMultiAnnotations() {
        rewriteRun(
          kotlin(
            """
              @Suppress("ALL")   @Deprecated("",ReplaceWith("Any()")) class Test {
              }
              """,
            """
              @Suppress("ALL")
                 @Deprecated("",ReplaceWith("Any()"))
               class Test {
              }
              """
          )
        );
    }

    @Test
    void annotatedClassDeclAlreadyCorrect() {
        rewriteRun(
          kotlin(
            """
              @Suppress("ALL")
              class Test {
              }
              """
          )
        );
    }

    @Test
    void annotatedClassDeclWithModifiers() {
        rewriteRun(
          kotlin(
            """
              @Suppress("ALL") public class Test {
              }
              """,
            """
              @Suppress("ALL")
               public class Test {
              }
              """
          )
        );
    }

    @Test
    void annotatedVariableDecl() {
        rewriteRun(
          kotlin(
            """
              class Test {
                  public fun doSomething() {
                      @Suppress("ALL") var foo: Int
                  }
              }
              """
          )
        );
    }

    @Test
    void annotatedVariableDeclWithModifier() {
        rewriteRun(
          kotlin(
            """
              class Test {
                  @Suppress("ALL")   private var foo: Int = 0
              }
              """,
            """
              class Test {
                  @Suppress("ALL")
                 private var foo: Int = 0
              }
              """
          )
        );
    }

    @Test
    void annotatedVariableDeclInMethodDeclaration() {
        rewriteRun(
          kotlin(
            """
              class Test {
                  public fun doSomething(@Suppress("ALL") foo: Int) {
                  }
              }
              """
          )
        );
    }

    @Issue("https://github.com/openrewrite/rewrite/issues/2469")
    @Test
    void elseOnNewLine() {
        rewriteRun(
          wrappingAndBraces(
            spaces -> spaces,
            wrap -> wrap.withIfStatement(IntelliJ.wrappingAndBraces().getIfStatement().withElseOnNewLine(true))),
          kotlin(
            """
              class Test {
                  fun method(arg0: Int) {
                      if (arg0 == 0) {
                          System.out.println("if");
                      } else if (arg0 == 1) {
                          System.out.println("else if");
                      } else {
                          System.out.println("else");
                      }
                  }
              }
              """,
            """
              class Test {
                  fun method(arg0: Int) {
                      if (arg0 == 0) {
                          System.out.println("if");
                      }
                      else if (arg0 == 1) {
                          System.out.println("else if");
                      }
                      else {
                          System.out.println("else");
                      }
                  }
              }
              """
          )
        );
    }

    @Issue("https://github.com/openrewrite/rewrite/issues/2469")
    @Test
    void elseNotOnNewLine() {
        rewriteRun(
          wrappingAndBraces(
            spaces -> spaces,
            wrap -> wrap.withIfStatement(IntelliJ.wrappingAndBraces().getIfStatement().withElseOnNewLine(false))),
          kotlin(
            """
              class Test {
                  fun method(arg0: Int) {
                      if (arg0 == 0) {
                          System.out.println("if");
                      }
                      else if (arg0 == 1) {
                          System.out.println("else if");
                      }
                      else {
                          System.out.println("else");
                      }
                  }
              }
              """,
            """
              class Test {
                  fun method(arg0: Int) {
                      if (arg0 == 0) {
                          System.out.println("if");
                      } else if (arg0 == 1) {
                          System.out.println("else if");
                      } else {
                          System.out.println("else");
                      }
                  }
              }
              """
          )
        );
    }

    @Issue("https://github.com/openrewrite/rewrite/issues/3191")
    @Test
    void emptyLineBeforeEnumConstants() {
        rewriteRun(
          kotlin(
            """
              enum class Status {
                  NOT_STARTED,
                  STARTED
              }
              """
          )
        );
    }

    @Test
    void singleStatementFunctionNoNewLines() {
        rewriteRun(
          kotlin(
            """
              class A {
                  fun name(): String =
                          "123"
              }
              """
          )
        );
    }

    @Test
    void nonSingleStatementFunctionNeedNewLines() {
        // An equivalent code with above test singleStatementFunctionNoNewLines, but not a single statement function
        rewriteRun(
          kotlin(
            """
              class A {
                  fun name(): String {   return "123" }
              }
              """,
            """
              class A {
                  fun name(): String {
                 return "123"
               }
              }
              """
          )
        );
    }

    @Test
    void emptyClassBody() {
        rewriteRun(
          kotlin(
            """
              class Test {}
              """
          )
        );
    }

    @Test
    void primaryConstructorAnnotations() {
        rewriteRun(
          kotlin(
            """
              class T @Suppress constructor()
              """
          )
        );
    }

    @Test
    void trailingAnnotationCommentBeforeClass() {
        rewriteRun(
          kotlin(
            """
              class T {
                  @Suppress // comment
                  class A
              }
              """
          )
        );
    }

    @Test
    void annotatedLocalVariable() {
        rewriteRun(
          kotlin(
            """
              fun f() {
                  @Suppress val i = 1
                  @Suppress val j = 1
              }

              val o = object {
                  @Suppress val i = 1
                  @Suppress val j = 1
              }

              class T {
                  @Suppress val i = 1
                  @Suppress val j = 1
              }
              """,
            """
              fun f() {
                  @Suppress val i = 1
                  @Suppress val j = 1
              }

              val o = object {
                  @Suppress
               val i = 1
                  @Suppress
               val j = 1
              }

              class T {
                  @Suppress
               val i = 1
                  @Suppress
               val j = 1
              }
              """
          )
        );
    }
}

File: rewrite-java-tck/src/main/kotlin/org/openrewrite/java/format/WrappingAndBracesTest.kt
/*
 * Copyright 2020 the original author or authors.
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * https://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.openrewrite.java.format

import org.junit.jupiter.api.Test
import org.openrewrite.Issue
import org.openrewrite.Tree
import org.openrewrite.java.Assertions.java
import org.openrewrite.java.JavaParser
import org.openrewrite.java.style.IntelliJ
import org.openrewrite.java.style.WrappingAndBracesStyle
import org.openrewrite.style.NamedStyles
import org.openrewrite.style.Style
import org.openrewrite.test.RecipeSpec
import org.openrewrite.test.RewriteTest
import org.openrewrite.test.RewriteTest.toRecipe

@Suppress("UnusedAssignment", "ClassInitializerMayBeStatic")
interface WrappingAndBracesTest : RewriteTest {

    override fun defaults(spec: RecipeSpec) {
        spec.recipe(toRecipe {WrappingAndBracesVisitor(WrappingAndBracesStyle(WrappingAndBracesStyle.IfStatement(false)))})
    }

    fun namedStyles(styles: Collection<Style>) : Iterable<NamedStyles> {
        return listOf(NamedStyles(Tree.randomId(), "Test", "test", "test", emptySet(), styles))
    }

    @Suppress("StatementWithEmptyBody", "ConstantConditions")
    @Issue("https://github.com/openrewrite/rewrite/issues/804")
    @Test
    fun conditionalsShouldStartOnNewLines() = rewriteRun(
        { spec -> spec.recipe(WrappingAndBraces().doNext(TabsAndIndents()))},
        java("""
            class Test {
                void test() {
                    if (1 == 2) {
                    } if (1 == 3) {
                    }
                }
            }
            """,
            """
            class Test {
                void test() {
                    if (1 == 2) {
                    }
                    if (1 == 3) {
                    }
                }
            }
            """
        )
    )

    @Test
    fun blockLevelStatements() = rewriteRun(
        java("""
            public class Test {
                {        int n = 0;
                    n++;
                }
            }
            """,
        """
            public class Test {
                {
                    int n = 0;
                    n++;
                }
            }
            """
        )
    )

    @Test
    fun blockEndOnOwnLine() = rewriteRun(
        java("""
            class Test {
                int n = 0;}
            """,
        """
            class Test {
                int n = 0;
            }
            """
        )
    )

    @Test
    fun annotatedMethod() = rewriteRun(
        java("""
            public class Test {
                @SuppressWarnings({"ALL"}) Object method() {
                    return new Object();
                }
            }
            """,
        """
            public class Test {
                @SuppressWarnings({"ALL"})
             Object method() {
                    return new Object();
                }
            }
            """
        )
    )

    @Test
    fun annotatedMethodWithModifier() = rewriteRun(
        java("""
            public class Test {
                @SuppressWarnings({"ALL"}) public Object method() {
                    return new Object();
                }
            }
            """,
        """
            public class Test {
                @SuppressWarnings({"ALL"})
             public Object method() {
                    return new Object();
                }
            }
            """
        )
    )

    @Test
    fun annotatedMethodWithModifiers() = rewriteRun(
        java("""
            public class Test {
                @SuppressWarnings({"ALL"}) public final Object method() {
                    return new Object();
                }
            }
            """,
        """
            public class Test {
                @SuppressWarnings({"ALL"})
             public final Object method() {
                    return new Object();
                }
            }
            """
        )
    )

    @Test
    fun annotatedMethodWithTypeParameter() = rewriteRun(
        java("""
            public class Test {
                @SuppressWarnings({"ALL"}) <T> T method() {
                    return null;
                }
            }
            """,
        """
            public class Test {
                @SuppressWarnings({"ALL"})
             <T> T method() {
                    return null;
                }
            }
            """
        )
    )

    @Test
    fun multipleAnnotatedMethod() = rewriteRun(
        java("""
            public class Test {
                @SuppressWarnings({"ALL"}) @Deprecated Object method() {
                    return new Object();
                }
            }
            """,
        """
            public class Test {
                @SuppressWarnings({"ALL"})
             @Deprecated
             Object method() {
                    return new Object();
                }
            }
            """
        )
    )

    @Test
    fun annotatedConstructor() = rewriteRun(
        java("""
            public class Test {
                @SuppressWarnings({"ALL"}) @Deprecated Test() {
                }
            }
            """,
        """
            public class Test {
                @SuppressWarnings({"ALL"})
             @Deprecated
             Test() {
                }
            }
            """
        )
    )

    @Test
    fun annotatedClassDecl() = rewriteRun(
        java("""
            @SuppressWarnings({"ALL"}) class Test {
            }
            """,
        """
            @SuppressWarnings({"ALL"})
             class Test {
            }
            """
        )
    )

    @Test
    fun annotatedClassDeclAlreadyCorrect() = rewriteRun(
        java("""
            @SuppressWarnings({"ALL"}) 
            class Test {
            }
            """
        )
    )

    @Test
    fun annotatedClassDeclWithModifiers() = rewriteRun(
        java("""
            @SuppressWarnings({"ALL"}) public class Test {
            }
            """,
        """
            @SuppressWarnings({"ALL"})
             public class Test {
            }
            """
        )
    )

    @Test
    fun annotatedVariableDecl() = rewriteRun(
        java("""
            public class Test {
                public void doSomething() {
                    @SuppressWarnings("ALL") int foo;        
                }
            }
            """,
        """
            public class Test {
                public void doSomething() {
                    @SuppressWarnings("ALL")
             int foo;        
                }
            }
            """
        )
    )

    @Test
    fun annotatedVariableAlreadyCorrect() = rewriteRun(
        java("""
            public class Test {
                public void doSomething() {
                    @SuppressWarnings("ALL")
                    int foo;        
                }
            }
            """
        )
    )

    @Test
    fun annotatedVariableDeclWithModifier() = rewriteRun(
        java("""
            public class Test {
                @SuppressWarnings("ALL") private int foo;        
            }
            """,
        """
            public class Test {
                @SuppressWarnings("ALL")
             private int foo;        
            }
            """
        )
    )

    @Test
    fun annotatedVariableDeclInMethodDeclaration() = rewriteRun(
        java("""
            public class Test {
                public void doSomething(@SuppressWarnings("ALL") int foo) {
                }
            }
            """
        )
    )

    @Issue("https://github.com/openrewrite/rewrite/issues/375")
    @Test
    fun retainTrailingComments() = rewriteRun(
        java("""
            public class Test {
            int m; /* comment */ int n;}
            """,
        """
            public class Test {
            int m; /* comment */
            int n;
            }
            """
        )
    )

    @Issue("https://github.com/openrewrite/rewrite/issues/2469")
    @Test
    fun elseOnNewLine() = rewriteRun(
        { spec ->
            spec.parser(
                JavaParser.fromJavaVersion().styles(
                    namedStyles(listOf(IntelliJ.spaces().run {
                        withBeforeKeywords(beforeKeywords.run {
                            withElseKeyword(true)
                        })
                    }))
                )
            ).recipe(toRecipe {WrappingAndBracesVisitor(WrappingAndBracesStyle(WrappingAndBracesStyle.IfStatement(true)))}
                .doNext(TabsAndIndents()))
        },
        java("""
            public class Test {
                void method(int arg0) {
                    if (arg0 == 0) {
                        System.out.println("if");
                    } else if (arg0 == 1) {
                        System.out.println("else if");
                    } else {
                        System.out.println("else");
                    }
                }
            }
            """,
            """
            public class Test {
                void method(int arg0) {
                    if (arg0 == 0) {
                        System.out.println("if");
                    }
                    else if (arg0 == 1) {
                        System.out.println("else if");
                    }
                    else {
                        System.out.println("else");
                    }
                }
            }
            """
        )
    )

    @Issue("https://github.com/openrewrite/rewrite/issues/2469")
    @Test
    fun elseNotOnNewLine() = rewriteRun(
        { spec ->
            spec.parser(
                JavaParser.fromJavaVersion().styles(
                    namedStyles(listOf(IntelliJ.spaces().run {
                        withBeforeKeywords(beforeKeywords.run {
                            withElseKeyword(true)
                        })
                    }))
                )
            ).recipe(toRecipe {WrappingAndBracesVisitor(WrappingAndBracesStyle(WrappingAndBracesStyle.IfStatement(false)))}
                .doNext(Spaces())
                .doNext(TabsAndIndents()))
        },
        java("""
            public class Test {
                void method(int arg0) {
                    if (arg0 == 0) {
                        System.out.println("if");
                    }
                    else if (arg0 == 1) {
                        System.out.println("else if");
                    }
                    else {
                        System.out.println("else");
                    }
                }
            }
            """,
            """
            public class Test {
                void method(int arg0) {
                    if (arg0 == 0) {
                        System.out.println("if");
                    } else if (arg0 == 1) {
                        System.out.println("else if");
                    } else {
                        System.out.println("else");
                    }
                }
            }
            """
        )
    )
}
